lib/external_service_new/shopify_downloader.rb

```diff
@@ -28,37 +29,122 @@ def each_email(id:, &block)
+      "province",
+      "province_code",
+      "zip"]
+
+    attr_reader :dispatcher, :logger
     def initialize(dispatcher:, filters:, logger:nil)
       raise "dispatcher must not be nil, found #{dispatcher.inspect}" unless dispatcher
 
-      @dispatcher = dispatcher
-      @filters    = filters
-      @logger     = logger || respond_to?(:logger) ? logger : nil
+      @dispatcher           = dispatcher
+      @remote_filters       = filters.select{|filter| REMOTE_FILTERS.include?(filter.fetch(:code))}
+      @after_remote_filters = filters.select{|filter| AFTER_REMOTE_FILTERS.include?(filter.fetch(:code))}
+      @logger               = logger || respond_to?(:logger) ? logger : nil
     end
 
+
+    EMAIL_FIELDS = "email,orders_count,total_spent,created_at,updated_at"
     def each_email(&block)
       return to_enum(:each_email) unless block
 
-      logger && logger.info("Downloading Shopify Customer List with #{filters ? filters.inspect : "no filters"}")
+      filter_log_message = (@remote_filters + @after_remote_filters).empty? ? 'no filters' : parse_filters(@remote_filters) + " " + parse_filters(@after_remote_filters)
+      logger && logger.info("Downloading Shopify Customer List with #{filter_log_message}")
 
       page = 1
       loop do
-        response = dispatcher.dispatch(:get, "/admin/customers/search.json?page=#{page}&limit=250&fields=email")
+        response = dispatcher.dispatch(:get, "/admin/customers/search.json?page=#{page}&limit=250&fields=#{EMAIL_FIELDS}&query= " + parse_filters(@remote_filters))
         break if response["customers"].empty?
 
-        response["customers"].each do |customer|
-          next unless customer && customer["email"]
+        response["customers"].each do |shopify_customer|
+          next unless matching_after_remote_filters(shopify_customer)
 
-          yield customer["email"]
+          yield shopify_customer["email"]
         end
         page += 1
       end
 
       logger && logger.info("Downloaded Shopify Customer List")
     end
 
+
     private
 
-    attr_reader :dispatcher, :filters, :logger
+    def parse_filters(filters)
+      filters.map{|f| "#{f.fetch(:code)}:#{f.fetch(:value)}"}.join(" ")
+    end
+
+    def matching_after_remote_filters(shopify_customer)
+      shopify_customer["total_spent"] = shopify_customer["total_spent"].to_i
+      shopify_customer["orders_count"] = shopify_customer["orders_count"].to_i
+      shopify_customer["created_at"] = shopify_customer["created_at"].to_time
+      shopify_customer["updated_at"] = shopify_customer["updated_at"].to_time
+
+      @after_remote_filters.all? do |filter|
+        self.send("check_#{filter.fetch(:code)}", shopify_customer, filter.fetch(:value))
+      end
+    end
+
+
+    # AFTER_REMOTE_FILTER CHECKS
+
+    def check_created_at_max(shopify_customer, time)
+      shopify_customer["created_at"] <= time
+    end
+
+    def check_created_at_min(shopify_customer, time)
+      shopify_customer["created_at"] >= time
```

Comment:
no need for the `shopify_` prefix, there is only one type of customer in this class.